End Biomes Api [1.16]

RedstoneParadox opened the pull request:
This PR adds an API which allows for adding biomes to one of the five "regions" of the End corresponding to the vanilla biomes. It uses a weight system similar to the Overworld biome API. It also does some internal changes, mainly to `WeightedBiomePicker`.

~~Would it be desirable to add custom regions to the End? What vanilla does is just pick base the biome off of a noise value from 40 to -40 (And a circle for the main island), but it's hard-coded so adding in custom regions/ranges is going to be more intrusive. EDIT: Should mention that doing this may cause noticeable deviations from vanilla end generation.~~

RedstoneParadox committed:
Create EndBiomes.java

RedstoneParadox committed:
Add the ability to add modded dimensions to the regions of the end.

RedstoneParadox committed:
Revert "Create EndBiomes.java"

This reverts commit 4d1736fad99e1be0921ef92d9576e8f503e64be4.

RedstoneParadox committed:
Revert "Revert "Create EndBiomes.java""

This reverts commit 673c508d825a0eb095f94ff9b5cb88d1b77e395c.

RedstoneParadox committed:
Rename ContinentalBiomeEntry to WeightedBiomeEntry

RedstoneParadox committed:
Complete API for adding biomes to the end

RedstoneParadox committed:
Mixin to end biome source

RedstoneParadox committed:
Revert "Rename ContinentalBiomeEntry to WeightedBiomeEntry"

This reverts commit 849197e15d18c26af2a6f2aa1766b7d73d9a7144.

RedstoneParadox committed:
Revert "Revert "Rename ContinentalBiomeEntry to WeightedBiomeEntry""

This reverts commit c2aa4ab097eb3daed71e82d4e77142671ed04316.

RedstoneParadox committed:
Create SimpleLayerRandomnessSource for use with end biome source mixin

RedstoneParadox committed:
Renamed MAIN region to MAIN_ISLANDS

RedstoneParadox committed:
Create test mod

RedstoneParadox committed:
Added pickFromNoise method to WeightedBiomePicker

RedstoneParadox committed:
Javadoc and licensing info

RedstoneParadox committed:
Comply with checkstyle

RedstoneParadox committed:
Internalize EndRegion

RedstoneParadox committed:
Added stretching when getting the biome for the end

i509VCB commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/EndBiomes.java:
```diff
@@ -0,0 +1,102 @@
+	 * heavier biomes being more likely to be selected and lighter biomes being selected with less likelihood.
+	 * Vanilla biomes have a weight of 1.0
+	 */
+	public static void addMidlandsiome(Biome biome, double weight) {
```
```suggestion
	public static void addMidlandsBiome(Biome biome, double weight) {
```

i509VCB commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/InternalBiomeData.java:
```diff
@@ -131,6 +133,13 @@ public static void addNetherBiome(Biome biome) {
+	public static void addEndBiome(Biome biome, EndRegion region, double weight) {
+		Preconditions.checkArgument(biome != null, "Biome is null");
+		Preconditions.checkArgument(region != null, "EndRegion is null");
+		Preconditions.checkArgument(weight > 0.0, "Weight is less than or equal to 0.0 (%s)", weight);
```
```suggestion
		Preconditions.checkArgument(weight > 0.0, "Weight is less than or equal to 0.0 (got %s)", weight);
```

i509VCB commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/InternalBiomeData.java:
```diff
@@ -131,6 +133,13 @@ public static void addNetherBiome(Biome biome) {
+		Preconditions.checkArgument(biome != null, "Biome is null");
+		Preconditions.checkArgument(region != null, "EndRegion is null");
+		Preconditions.checkArgument(weight > 0.0, "Weight is less than or equal to 0.0 (%s)", weight);
+		END_REGION_BIOME_PICKERS.computeIfAbsent(region, region1 -> new WeightedBiomePicker()).addBiome(biome, weight);
```
```suggestion
		END_REGION_BIOME_PICKERS.computeIfAbsent(region, key -> new WeightedBiomePicker()).addBiome(biome, weight);
```

i509VCB commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheEndBiomeSource.java:
```diff
@@ -0,0 +1,74 @@
+	@Shadow
+	@Final
+	private long field_24731;
+	private final Map<Biome, EndRegion> BIOME_REGION_MAP = new HashMap<>();
```
Mark this `@Unique`

i509VCB commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheEndBiomeSource.java:
```diff
@@ -0,0 +1,74 @@
+	@Final
+	private long field_24731;
+	private final Map<Biome, EndRegion> BIOME_REGION_MAP = new HashMap<>();
+	private final LayerRandomnessSource randomnessSource = new SimpleLayerRandomnessSource(field_24731);
```
Mark this `@Unique`

i509VCB commented on fabric-biomes-v1/src/testmod/java/net/fabricmc/fabric/test/biome/FabricBiomeTest.java:
```diff
@@ -31,8 +33,15 @@
 	public class TestCrimsonForestBiome extends CrimsonForestBiome {
 	}
+
+	public class TestEndHighlandsBiome extends EndHighlandsBiome {
+
+	}
```
```suggestion
	public class TestEndHighlandsBiome extends EndHighlandsBiome {
	}
```

So checkstyle builds

RedstoneParadox replied:
That doesn't prevent it from building.

i509VCB replied:
must have been mistaken. I'd say then to be consistent with the other test class

i509VCB reviewed (commented):
I don't know much about biomes tbf (hence why I requested supercoder for review). Most of my feedback is related to documentation and method names.

Prospector commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/api/biomes/v1/EndBiomes.java:
```diff
@@ -0,0 +1,102 @@
+/**
+ * API that allows for adding biomes to the biome source for The End.
+ */
+public final class EndBiomes {
```
TheEndBiomes imo

RedstoneParadox committed:
Fix typo

Courtesy of i509VBC

Co-authored-by: i509VCB <i509vcb@gmail.com>

RedstoneParadox committed:
Rename EndBiomes to TheEndBiomes to be consistent with yarn mappings

RedstoneParadox committed:
Made end test biome consistent in style with previous test

Courtesy of i509VBC

Co-authored-by: i509VCB <i509vcb@gmail.com>

RedstoneParadox committed:
Add @Unique to private fields in MixinTheEndBiomeSource

Prospector commented:
The checkstyle *is* failing though, glancing at the log it looked like import order was one of the issues.

RedstoneParadox committed:
Fix imports

RedstoneParadox committed:
Didn't know test mods needed to follow the checkstyle

liach commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/InternalBiomeData.java:
```diff
@@ -51,6 +51,8 @@ private InternalBiomeData() { }
 
 	private static final Set<Biome> NETHER_BIOMES = new HashSet<>();
 
+	private static final Map<EndRegion, WeightedBiomePicker> END_REGION_BIOME_PICKERS = new HashMap<>();
```
```suggestion
	private static final Map<EndRegion, WeightedBiomePicker> END_REGION_BIOME_PICKERS = new EnumMap<>();
```

liach replied:
Don't just apply this change, you need to add import in addition

liach commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/impl/biome/SimpleLayerRandomnessSource.java:
```diff
@@ -0,0 +1,44 @@
+	public SimpleLayerRandomnessSource(long seed) {
+		this.seed = seed;
+		this.random = new Random(seed);
+		this.sampler = new PerlinNoiseSampler(new Random(seed));
```
Is it intentional that the calls on the sampler doesn't affect the calls on nextInt in this implementation?

liach commented on fabric-biomes-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheEndBiomeSource.java:
```diff
@@ -0,0 +1,77 @@
+
+		if (BIOME_REGION_MAP.containsKey(vanillaBiome)) {
+			EndRegion region = BIOME_REGION_MAP.get(vanillaBiome);
+			WeightedBiomePicker picker = InternalBiomeData.getEndRegionBiomePickers().get(region);
```
Maybe comment that `picker` won't be `null` as you've populated all the 5 end regions above, so if npe arises below it'd be easier to debug

RedstoneParadox committed:
SimpleLayerRandomnessSource's random is now use to instantiate its noise sampler

RedstoneParadox committed:
A very important note

RedstoneParadox committed:
Update javadoc for adding biomes to the Main Island region

RedstoneParadox committed:
Remove method that is leftover from when EndRegion was part of the API

RedstoneParadox committed:
Merge branch '1.16' of https://github.com/FabricMC/fabric into 1.16

ghost commented:
This should really be closed and discussion should move over to #1036 as adding end biomes is one of the goals of the new issue.

RedstoneParadox commented:
> This should really be closed and discussion should move over to #1036 as adding end biomes is one of the goals of the new issue.

According to that issue, adding End Biomes is probably not happening, so unless that changes, it's best to keep this open.

shartte committed:
Ported 1.16.1 biomes-api-v1 to 1.16.2.
Public API changes:
- Removed flagging biomes as suitable for player spawns since that is now handled in the Biome Builder.
- Changed API over to RegistryKey's, not because it's necessary, but because it is more ergonomic since Vanilla Biomes in BuiltInBiomes are now all exposed as keys, rather than instances.

shartte commented:
@RedstoneParadox I ported the 1.16.1 biome-v1 API to 1.16.2, and went ahead and rebased your API on top and also ported it to 1.16.2. Since I don't think  I can PR a force-push to your branch, please take a look at https://github.com/shartte/fabric/tree/ender-biomes-api and if you find that (and my changes to port it to 1.16.2) suitable, you might want to reset your branch to it, and update this PR. This will however make it depend on #1053

shartte committed:
Increase memory size to fix build failures.

RedstoneParadox commented:
Merged the changes made by @shartte but I don't know if it actually runs; I don't have the time to check that right now and I'll need to do it later.

i509VCB commented:
Needs to be merged with current 1.16 branch before reviews can continue

RedstoneParadox committed:
Merge branch '1.16' of https://github.com/FabricMC/fabric into 1.16

RedstoneParadox commented:
Rebased it, but I have to make sure everything is sorted out first.

RedstoneParadox committed:
Hopefully fixed everything that went wrong during the rebase

RedstoneParadox committed:
Update fabric-biome-api-v1.mixins.json

RedstoneParadox committed:
Finally got custom end biomes to generate

RedstoneParadox commented:
Something I just thought of is that modders may want certain midlands and lowlands biomes to be associated with their highlands biome since they're sort of equivalent to beaches in the overworld.

RedstoneParadox committed:
Thanks for nothing, GitHub Desktop

RedstoneParadox committed:
Bad GitHub Desktop

RedstoneParadox committed:
Bad newline no cookie

RedstoneParadox committed:
Got everything to follow the checkstyle

RedstoneParadox commented:
~~I've also noticed that the method I'm using to actually pick which biome replaces the vanilla one is not the best as it picks at random.~~

Edit: Disregard.

RedstoneParadox committed:
Bring branch up-to-date with main repo (#11)

OroArmor commented on fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/impl/biome/InternalBiomeData.java:
```diff
@@ -142,6 +161,13 @@ public static void addNetherBiome(RegistryKey<Biome> biome, Biome.MixedNoisePoin
 
+	public static void addEndBiomeReplacement(RegistryKey<Biome> replaced, RegistryKey<Biome> variant, double weight) {
+		Preconditions.checkNotNull(replaced, "replaced is null");
+		Preconditions.checkNotNull(variant, "replaced is null");
```
this should be `variant is null`

OroArmor commented on fabric-item-api-v1/src/testmod/java/net/fabricmc/fabric/test/item/FabricItemSettingsTests.java:
```diff
@@ -16,20 +16,31 @@
+import net.minecraft.world.biome.CrimsonForestBiome;
+import net.minecraft.world.biome.EndHighlandsBiome;
 
+import net.fabricmc.fabric.api.biomes.v1.TheEndBiomes;
 import net.fabricmc.api.ModInitializer;
-import net.fabricmc.fabric.api.item.v1.FabricItemSettings;
-
-public class FabricItemSettingsTests implements ModInitializer {
-	@Override
-	public void onInitialize() {
-		// Registers an item with a custom equipment slot.
-		Item testItem = new Item(new FabricItemSettings().group(ItemGroup.MISC).equipmentSlot(stack -> EquipmentSlot.CHEST));
-		Registry.register(Registry.ITEM, new Identifier("fabric-item-api-v1-testmod", "test_item"), testItem);
+import net.fabricmc.fabric.api.biomes.v1.NetherBiomes;
+
+public class FabricBiomeTest implements ModInitializer {
+	public static final String MOD_ID = "fabric-biome-api-v1-testmod";
+
+	@Override public void onInitialize() {
+		TestCrimsonForestBiome biome = Registry.register(Registry.BIOME, new Identifier(MOD_ID, "test_crimson_forest"), new TestCrimsonForestBiome());
+		NetherBiomes.addNetherBiome(Biomes.BEACH);
+		NetherBiomes.addNetherBiome(biome);
+
+		TestEndHighlandsBiome testEndBiome = Registry.register(Registry.BIOME, new Identifier(MOD_ID, "test_end_highlands"), new TestEndHighlandsBiome());
+		TheEndBiomes.addHighlandsBiome(testEndBiome, 5.0);
+	}
+
+	public class TestCrimsonForestBiome extends CrimsonForestBiome {
+	}
+
+	public class TestEndHighlandsBiome extends EndHighlandsBiome {
 	}
 }
```
uh is git being weird... This is in the wrong module, It should be in the biomes module, but it is in the Items module.

OroArmor reviewed (changes requested):
Looks ok, but you are including a bunch of other prs and other code that has nothing to do with this pr. This should cleaned up for your otherwise great pr to be considered to be merged in. There are stuff from villagers, tristate, tools, fabric renderer, and more that have nothing to do with end biomes.

RedstoneParadox committed:
Revert "Bring branch up-to-date with main repo (#11)"

This reverts commit dc471062410ec2de0899a7520584b16f5ddb88f1.

RedstoneParadox committed:
Remove some stuff that's not supposed to be there

RedstoneParadox committed:
Fix git silliness

RedstoneParadox committed:
Merge pull request #12 from FabricMC/1.16

Merge recent Fabric 1.16 commits, take 2

RedstoneParadox committed:
Update FabricBiomeTest.java

RedstoneParadox committed:
TheEndBiomes is now Deprecated to match OverworldBiomes and NetherBiomes

RedstoneParadox committed:
Update some checkNotNull messages

RedstoneParadox committed:
Change the noise scale when replacing end biomes

RedstoneParadox committed:
Bad space no cookie

RedstoneParadox committed:
Remove unnecessary code

RedstoneParadox committed:
Remove unused imports

OroArmor commented:
Wow its so much cleaner thanks. Looks great my dude.

i509VCB reviewed (approved):
From the api and implementation everything looks good, however I do want to consider the idea mentioned about border-like biomes that was brought up.

RedstoneParadox commented:
I think I might just go ahead and change the API to treat the Midlands and Barrens biomes as border-like biomes since no-one has voiced any sort of disapproval.

RedstoneParadox committed:
Set up InternalBiomeData to treat End Midlands and End Barrens as border biomes

RedstoneParadox committed:
Changed the API to reflect midlands and barrens biomes being considered border biomes

RedstoneParadox committed:
Start work on getting the new system fully working

RedstoneParadox committed:
Finally got everything working and cleaned up the Javadoc

RedstoneParadox committed:
Fixed checkstyle violations

RedstoneParadox committed:
Merge branch '1.16' into end-biomes-api

RedstoneParadox committed:
Fix checkstyle violations again

i509VCB commented on fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheEndBiomeSource.java:
```diff
@@ -0,0 +1,81 @@
+	private LayerRandomnessSource randomnessSource = new SimpleLayerRandomnessSource(seed);
+
+	@Inject(method = "getBiomeForNoiseGen", at = @At("RETURN"), cancellable = true)
+	private void fabric_getWeightedEndBiome(int biomeX, int biomeY, int biomeZ, CallbackInfoReturnable<Biome> cir) {
```
the `fabric_` prefix can be dropped

i509VCB reviewed (approved):
Everything looks good to me.

RedstoneParadox committed:
Drop fabric_ prefix

Player3324 reviewed (approved):
LGTM, thanks!

RedstoneParadox commented on fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/mixin/biome/MixinTheEndBiomeSource.java:
```diff
@@ -0,0 +1,81 @@
+		} else {
+			// Since the pickers are statically populated by InternalBiomeData, picker will never be null.
+			WeightedBiomePicker picker = InternalBiomeData.getEndBiomesMap().get(vanillaKey);
+			replacementKey = picker.pickFromNoise(randomnessSource, biomeX/64.0, 0, biomeZ/64.0);
```
One thing that was never actually addressed was whether or not this stretch value of 64 was good for biome sampling. Since `TheEndBiomeSource` is very simplistic, there isn't actually anything to base this value off of and was something I came up with. Perhaps 128 or 256 would be better?

RedstoneParadox replied:
I did some testing and it's very difficult to find a custom biome when the stretch is 128, so I'm just going to leave it as is.

i509VCB replied:
It would be nice if you throw a comment there explaining why 64 is the best value for say

jaskarth commented on fabric-biome-api-v1/src/main/java/net/fabricmc/fabric/impl/biome/SimpleLayerRandomnessSource.java:
```diff
@@ -0,0 +1,44 @@
+
+	public SimpleLayerRandomnessSource(long seed) {
+		this.seed = seed;
+		this.random = new Random(seed);
```
Don't use Java's random here. It'll cause issues. Instead, the layer random should be re-initialized with the x/z coords and world seed. (nextInt() is not used anywhere here, but it's important enough to point out. Alternatively, throw an unsupported operation exception in nextInt.)

RedstoneParadox replied:
But then what do I pass to the PerlinNoiseSampler?

RedstoneParadox committed:
Calling SimpleLayerRandomnessSource#nextInt() now throws an exception
